src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,464 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+using AsyncCompletionData = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using FilterResult = Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Session.FilterResult;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using Session = Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Session;
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.AsyncCompletion
+{
+    internal class ItemManager : IAsyncCompletionItemManager
+    {
+        private readonly CompletionHelper _completionHelper;
+
+        private const int MaxMRUSize = 10;
+        private ImmutableArray<string> _recentItems = ImmutableArray<string>.Empty;
+        private object _mruUpdateLock = new object();
+
+        internal ItemManager(IAsyncCompletionBroker broker)
+        {
+            // Session is created later and this is supported with CompletionTriggered.
+            // Handle it to subscribe for session events.
+            broker.CompletionTriggered += CompletionTriggered;
```

Comment:
When is this ever unsubscribed?

Reply:
removed events

---
In reply to: [234841015](https://github.com/dotnet/roslyn/pull/29016#discussion_r234841015) [](ancestors = 234841015)